cpp/src/fil/treelite_import.cu

```diff
@@ -616,129 +542,119 @@ void tl2fil_common(forest_params_t* params,
+    roots_.reserve(num_trees + 1);
+    roots_.push_back(0);
+    for (size_t i = 0; i < num_trees; ++i) {
+      int num_nodes = node_traits<fil_node_t>::IS_DENSE ? tree_num_nodes(params_.depth)
+                                                        : model_.trees[i].num_nodes;
+      roots_.push_back(num_nodes + roots_.back());
+    }
+    size_t total_nodes = roots_.back();
+    roots_.pop_back();
+
+    if (params_.leaf_algo == VECTOR_LEAF) {
+      size_t max_leaves = node_traits<fil_node_t>::IS_DENSE
+                            ? num_trees * (tree_num_nodes(params_.depth) + 1) / 2
+                            : (total_nodes + num_trees) / 2;
+      vector_leaf_.resize(max_leaves * params_.num_classes);
+    }
 
-  size_t num_trees = model.trees.size();
+    cat_sets_ = allocate_cat_sets_owner(model_);
+    nodes_.resize(total_nodes);
 
-  ptrees->reserve(num_trees);
-  ptrees->push_back(0);
-  for (size_t i = 0; i < num_trees - 1; ++i) {
-    ptrees->push_back(model.trees[i].num_nodes + ptrees->back());
-  }
-  size_t total_nodes = ptrees->back() + model.trees.back().num_nodes;
+// convert the nodes_
+#pragma omp parallel for
+    for (std::size_t tree_idx = 0; tree_idx < num_trees; ++tree_idx) {
+      // Max number of leaves processed so far
+      size_t leaf_counter = (roots_[tree_idx] + tree_idx) / 2;
+      tree2fil(nodes_,
+               roots_[tree_idx],
+               model_.trees[tree_idx],
+               tree_idx,
+               params_,
+               &vector_leaf_,
+               &leaf_counter,
+               &cat_sets_);
+    }
 
-  if (params->leaf_algo == VECTOR_LEAF) {
-    size_t max_leaves = (total_nodes + num_trees) / 2;
-    vector_leaf->resize(max_leaves * params->num_classes);
+    params_.num_nodes = nodes_.size();
   }
 
-  *cat_sets = allocate_cat_sets_owner(model);
-  pnodes->resize(total_nodes);
-
-// convert the nodes
-#pragma omp parallel for
-  for (std::size_t i = 0; i < num_trees; ++i) {
-    // Max number of leaves processed so far
-    size_t leaf_counter = ((*ptrees)[i] + i) / 2;
-    tree2fil_sparse(
-      *pnodes, (*ptrees)[i], model.trees[i], i, *params, vector_leaf, &leaf_counter, cat_sets);
+  /// initializes FIL forest object, to be ready to infer
+  void init_forest(const raft::handle_t& handle, forest_t* pforest, storage_type_t storage_type)
+  {
+    ML::fil::init(
+      handle, pforest, cat_sets_.accessor(), vector_leaf_, roots_.data(), nodes_.data(), &params_);
+    // sync is necessary as nodes_ are used in init(),
+    // but destructed at the end of this function
+    CUDA_CHECK(cudaStreamSynchronize(handle.get_stream()));
+    if (tl_params_.pforest_shape_str) {
+      *tl_params_.pforest_shape_str =
+        sprintf_shape(model_, storage_type, nodes_, roots_, cat_sets_);
+    }
   }
+};
 
-  params->num_nodes = pnodes->size();
+template <typename fil_node_t, typename threshold_t, typename leaf_t>
+void convert(const tl::ModelImpl<threshold_t, leaf_t>& model,
```

Comment:
It would be better if the parameter order here resembles that of `from_treelite`.